src/main/java/ch/njol/skript/expressions/ExprMaxStack.java

```diff
@@ -1,62 +1,126 @@
+	"Since MC 1.20.5 onwards, the maximum stack size of items can be changed (any integer from 1 to 99), and the said item can be stacked up to the set value, up to the maximum stack size of an inventory."
+})
+@Examples({
+	"send \"You can only pick up %max stack size of player's tool% of %type of (player's tool)%\" to player",
+	"set the maximum stack size of inventory of all players to 16",
+	"add 8 to the maximum stack size of player's tool",
+	"reset the maximum stack size of {_gui}"
+})
+@Since("2.1, INSERT VERSION (changeable, inventory support)")
+@RequiredPlugins("Spigot 1.20.5+ (changeable item max stack size)")
+public class ExprMaxStack extends SimplePropertyExpression<Object, Integer> {
 
 	static {
-		register(ExprMaxStack.class, Long.class, "max[imum] stack[[ ]size]", "itemtype");
+		register(ExprMaxStack.class, Integer.class, "max[imum] stack[[ ]size]", "itemtypes/inventories");
 	}
-	
-	@SuppressWarnings("null")
+
+	private static final boolean CHANGEABLE_ITEM_STACK_SIZE = Skript.methodExists(ItemMeta.class, "setMaxStackSize", Integer.class);
+
 	@Override
-	public Long convert(ItemType itemType) {
-		Object random = itemType.getRandomStackOrMaterial();
-		if (random instanceof Material)
-			return (long) ((Material) random).getMaxStackSize();
-		return (long) ((ItemStack) random).getMaxStackSize();
+	@Nullable
+	public Integer convert(Object from) {
+		if (from instanceof ItemType itemType)
+			return getMaxStackSize(itemType);
+		if (from instanceof Inventory inventory)
+			return inventory.getMaxStackSize();
+		return null;
 	}
 
 	@Override
-	public Class<? extends Long> getReturnType() {
-		return Long.class;
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+        return switch (mode) {
+            case ADD, REMOVE, RESET, SET -> {
+                if (!CHANGEABLE_ITEM_STACK_SIZE && ItemType.class.isAssignableFrom(getExpr().getReturnType())) {
+                    Skript.error("Changing the maximum stack size of items requires Minecraft 1.20.5 or newer!");
+                    yield null;
+                }
+                yield CollectionUtils.array(Integer.class);
+            }
+            default -> null;
+        };
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		for (Object source : getExpr().getArray(event)) {
+			Integer change = null;
+			if (mode != ChangeMode.RESET)
+				change = (int) delta[0];
+			if (source instanceof ItemType itemType) {
+				if (!CHANGEABLE_ITEM_STACK_SIZE)
+					continue;
+				int size = getMaxStackSize(itemType);
+                switch (mode) {
+                    case ADD:
+						size += change;
+                        break;
+                    case SET:
+						size = change;
+						break;
+                    case REMOVE:
+						size -= change;
+                        break;
+                }
```

Comment:
This can be an enhanced switch, like the one below with the `->` stuff

Reply:
Oops how did I missed out this old switch…